container/src/state.rs

```diff
@@ -0,0 +1,214 @@
+    Created,
+    #[serde(rename = "running")]
+    Running,
+    #[serde(rename = "stopped")]
+    Stopped,
+}
+
+impl Default for Status {
+    fn default() -> Self {
+        Status::Creating
+    }
+}
+
+/// Represent the state of the running container.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ContainerState {
+    id: String,
+    /// OCI version.
+    oci_version: String,
+    /// Runtime state of the container.
+    pub status: Arc<RwLock<Status>>,
+    /// ID of the container process.
+    pub pid: i32,
+    /// Path to the bundle.
+    bundle: PathBuf,
+}
+
+impl Default for ContainerState {
+    fn default() -> Self {
+        ContainerState {
+            oci_version: OCI_VERSION.to_string(),
+            id: String::default(),
+            status: Arc::new(RwLock::new(Status::default())),
+            pid: 0,
+            bundle: PathBuf::default(),
+        }
+    }
+}
+
+impl ContainerState {
+    pub fn new(id: &str, bundle_path: &str) -> Result<Self> {
+        ContainerState::_new(id, bundle_path, KAPS_ROOT_PATH)
+    }
+
+    fn _new(id: &str, bundle_path: &str, container_dir: &str) -> Result<Self> {
+        let bundle = PathBuf::from(bundle_path);
+        let container_path = PathBuf::from(container_dir).join(id);
+
+        if container_path.as_path().exists() {
+            return Err(Error::ContainerExists(format!(
+                "A container with the id '{}' already exists",
+                id
+            )));
+        }
+
+        // create the container directory
+        fs::create_dir_all(&container_path).map_err(Error::CreateStateFile)?;
+
+        // create the `state.json` file of the container
+        File::create(container_path.join(STATE_FILE)).map_err(Error::CreateStateFile)?;
+
+        let container_state = ContainerState {
+            id: id.to_string(),
+            bundle,
+            ..Default::default()
+        };
+
+        container_state.save(container_dir)?;
+
+        Ok(container_state)
+    }
+
+    /// Get the current runtime status of the container.
+    ///
+    /// As The container status is a RwLock,
+    /// calling this function results in acquiring a read lock on the status.
+    fn _status(&self) -> Result<Status> {
+        let container_status = Arc::clone(&self.status);
+        let container_status = container_status
+            .read()
+            .map_err(|e| Error::StatusLockPoisoned(e.to_string()))?;
+
+        Ok(*container_status)
+    }
+
+    /// Save the container state.
+    ///
+    /// The container state file must already have been created.
+    fn save(&self, container_dir: &str) -> Result<()> {
+        let container_path = PathBuf::from(container_dir).join(&self.id);
+
+        let file = OpenOptions::new()
+            .write(true)
+            .truncate(true)
+            .open(container_path.join(STATE_FILE))
+            .map_err(Error::OpenStateFile)?;
+
+        serde_json::to_writer_pretty(file, &self).map_err(Error::WriteStateFile)?;
```

Comment:
Since you're going to rebase that PR once more (CI is failing), you could clean that one further. Remove the `Ok(())` below and just do `serde_json::to_writer_pretty(file, &self).map_err(Error::WriteStateFile)`

Reply:
Indeed I was watching why the CI is falling and the error come from line 176 of `container/src/lib.rs`, which is weird because this line does not exist in the last version of my PR.
I have already tried to re-run the job but nothing changed.
Is there any way to clean the runner state?

Reply:
The runner state is clean. The problem is that you have not rebased your PR on top of the latest main branch. We merged PR https://github.com/virt-do/kaps/pull/21 and it introduces new unit tests.

You need to rebase and fix the new unit tests to follow your new APIs.

As a side note, if you want to check how your PR really looks like:

```bash
git fetch origin pull/20/merge:pr20
git checkout pr20
```

Reply:
Thanks for the tip @sameo I didn't know that the state of the PR can differ from my branch :open_mouth:.

All test is ok.
There is only one comment left above:  

> As the name of the container must be unique to the host I was thinking that `id` was best to describe this argument.

Reply:
> Thanks for the tip @sameo I didn't know that the state of the PR can differ from my branch :open_mouth:.
> 
> 
> 
> All test is ok.
> 
> There is only one comment left above:  
> 
> 
> 
> > As the name of the container must be unique to the host I was thinking that `id` was best to describe this argument.

2 things:

1. You call it a name in your option description.
2. You can not ask the user to provide an if that's unique, it's not a very good UX.

It should be a name, a friendly one, and it's your code responsibility to check that it's unique.

Reply:
Ok, I see.
My code already takes into considération the vérification, I'm coming to make the change for the argument.